go/src/lib.rs

```diff
@@ -1,9 +1,159 @@
+    channel_address: usize,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
+///
+/// This struct should be freed using `free_connection_response` to avoid memory leaks.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+}
+
+/// The glide client.
+// TODO: Remove allow(dead_code) once command logic is implemented
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+struct CreateClientError {
+    message: String,
+    error_type: RequestErrorType,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, CreateClientError> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Unspecified,
+        })?;
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let redis_error = err.into();
+            CreateClientError {
+                message: errors::error_message(&redis_error),
+                error_type: errors::error_type(&redis_error),
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime
+        .block_on(GlideClient::new(request))
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Disconnect,
+        })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+///
+/// The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes.
+/// * `connection_request_len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
```

Comment:
not high priority, but it would be nice if this function was async, too.

Reply:
Yes.  It would be nice.  Go doesn't really offer async capabilities very well.  Maybe a 'nice to have' for later on?

Reply:
@aaron-congo can you add a TODO to consider this later?

Reply:
I opened this PR for Jonathan so he's the one working on it, in hindsight it probably would've been clearer who owns the PR if he opened this, looks like he added a TODO here already though